### PR TITLE
feat: Make use of CSS layers

### PR DESCRIPTION
**Pull request by JoCa96**
Relates to #422, #338

- Created component mixin that applies component layer and normalize CSS
- Restructured styles and split them into global and mixins. This avoid repeated includes of global styles.
- Applied mixin for all components and removed redundant styles
- Updated screenshots as the fieldset had an unintentional margin

Added developer documentation in https://github.com/SchwarzIT/onyx/wiki/Contribution-Coding-Guidelines#css-layers

### CODE REVIEW

File: packages/sit-onyx/src/components/OnyxAppLayout/OnyxAppLayout.vue [hunk 1]
```diff
@@ -33,55 +33,57 @@ const slots = defineSlots<{
 </template>
 
 <style lang="scss">
+@use "../../styles/mixins/layers.scss";
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxAppLayout/OnyxAppLayout.vue:36]:
```suggestion
@use "@/styles/mixins/layers.scss";
```

(optional) Its maybe easier to import it like so :) Same for all the other components then

File: packages/sit-onyx/src/components/OnyxIconButton/OnyxIconButton.vue [hunk 2]
```diff
@@ -25,7 +25,7 @@ const emit = defineEmits<{
 
 <template>
   <button
-    class="onyx-icon-button"
+    class="onyx-icon-button onyx-reset"
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxIconButton/OnyxIconButton.vue:28]: Why is this needed?

File: packages/sit-onyx/src/styles/globals/density.scss
Comment (larsrickert) [inline packages/sit-onyx/src/styles/globals/density.scss:1]: Would prefer to remove the `globals` styles since all our styles are global anyways so we can keep the folder structure flat

File: packages/sit-onyx/src/styles/globals/layers.scss [hunk 3]
```diff
@@ -0,0 +1 @@
+@layer onyx.reset, onyx.utility, onyx.component, onyx.density;
```
Comment (larsrickert) [inline packages/sit-onyx/src/styles/globals/layers.scss:1]: What does this do / why is this needed? Maybe you can add a small comment

File: packages/sit-onyx/src/styles/globals/root.scss [hunk 4]
```diff
@@ -0,0 +1,30 @@
+@layer onyx.utility {
```
Comment (larsrickert) [inline packages/sit-onyx/src/styles/globals/root.scss:1]: Instead of putting all these general styles in a `utility` layer, can't we define a general `onyx` layer instead?

File: packages/sit-onyx/src/styles/mixins/layers.scss [hunk 5]
```diff
@@ -0,0 +1,11 @@
+@use "normalize.scss";
+
+@mixin component() {
+  @layer onyx.reset {
+    @include normalize.apply;
```
Comment (larsrickert) [inline packages/sit-onyx/src/styles/mixins/layers.scss:5]: Alternatively we could place the normalize styles itself in `onyx.reset` layer so we could remove this block and just define the below layers as `@layer onyx.reset onyx.component`

File: packages/sit-onyx/src/styles/mixins/normalize.scss [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  }
+
+  &,
+  * {
+    &,
+    &::before,
+    &::after {
+      margin: 0;
+
+      box-sizing: border-box;
```
Comment (larsrickert) [inline packages/sit-onyx/src/styles/mixins/normalize.scss:13]: So we can remove this from all our existing components, right?

File: packages/sit-onyx/src/styles/mixins/normalize.scss [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+      -moz-text-size-adjust: none;
+      -webkit-text-size-adjust: none;
+      text-size-adjust: none;
+    }
+
+    &:is(ul, ol) {
+      list-style: none;
```
Comment (larsrickert) [inline packages/sit-onyx/src/styles/mixins/normalize.scss:21]: Not sure if this is intended...what if I want to place a simple list in my project inside the app layout?

File: packages/sit-onyx/src/styles/mixins/normalize.scss [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    ================
+    */
+
+    /**
+    1. Add the correct height in Firefox.
+    2. Correct the inheritance of border color in Firefox. (https://bugzilla.mozilla.org/show_bug.cgi?id=190655)
+    */
+
+    &:is(hr) {
+      height: 0; /* 1 */
```
Comment (larsrickert) [inline packages/sit-onyx/src/styles/mixins/normalize.scss:41]: This will hide the element, right? Why would we want this?

File: packages/sit-onyx/src/styles/mixins/normalize.scss [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+      border-style: none;
+      padding: 0;
+    }
+
+    /**
+    Restore the focus styles unset by the previous rule.
+    */
+
+    &:-moz-focusring {
+      outline: 1px dotted ButtonText;
```
Comment (larsrickert) [inline packages/sit-onyx/src/styles/mixins/normalize.scss:143]:
```suggestion
      outline: var(--onyx-1px-in-rem) dotted ButtonText;
```

Also what is this magic `ButtonText` value?

**Review by larsrickert — CHANGES_REQUESTED** [review]

**Comment by larsrickert** [comment]
The bundle sizes is increasing by around 100kb. Is it possible that all the layers styles are redundantly copied to every component?

File: packages/sit-onyx/src/components/OnyxIconButton/OnyxIconButton.vue (hunk 2 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxIconButton/OnyxIconButton.vue:28]: Removed it. My first approach was to set a class.

File: packages/sit-onyx/src/components/OnyxAppLayout/OnyxAppLayout.vue (hunk 1 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxAppLayout/OnyxAppLayout.vue]: Done

File: packages/sit-onyx/src/styles/globals/density.scss
Comment (JoCa96) [reply packages/sit-onyx/src/styles/globals/density.scss:1]: If we don't separate them all globals will be duplicated with each include

File: packages/sit-onyx/src/styles/globals/layers.scss (hunk 3 above)
Comment (JoCa96) [reply packages/sit-onyx/src/styles/globals/layers.scss:2]: done

File: packages/sit-onyx/src/styles/globals/root.scss (hunk 4 above)
Comment (JoCa96) [reply packages/sit-onyx/src/styles/globals/root.scss:1]: See in [packages/sit-onyx/src/styles/globals/layers.scss](https://github.com/SchwarzIT/onyx/pull/768/files/20d7fb6c28eb4bdb8a515bd85c4f7a98ff6fe7c5#diff-7d8e57f01d19f241269bebb9d98720661ce8bdb62b45d21d46e28c6dff96cf28) we only define sub-layers.

File: packages/sit-onyx/src/styles/mixins/normalize.scss (hunk 6 above)
Comment (JoCa96) [reply packages/sit-onyx/src/styles/mixins/normalize.scss:13]: correct!

File: packages/sit-onyx/src/styles/mixins/normalize.scss (hunk 8 above)
Comment (JoCa96) [reply packages/sit-onyx/src/styles/mixins/normalize.scss:41]: updated comment

File: packages/sit-onyx/src/styles/mixins/layers.scss (hunk 5 above)
Comment (JoCa96) [reply packages/sit-onyx/src/styles/mixins/layers.scss:5]: thats not how layers work, layers only define the cascade not any rules

File: packages/sit-onyx/src/styles/mixins/normalize.scss (hunk 9 above)
Comment (JoCa96) [reply packages/sit-onyx/src/styles/mixins/normalize.scss:143]: A firefox specifc style value, as described in the comment we are setting the original default values here.

File: packages/sit-onyx/src/styles/globals/density.scss
Comment (larsrickert) [reply packages/sit-onyx/src/styles/globals/density.scss:1]: I mean we can still separate the files but move them 1 folder up so we don't need the globals folder

**Comment by JoCa96** [comment]
> The bundle sizes is increasing by around 100kb. Is it possible that all the layers styles are redundantly copied to every component?

Yes, the include happens for every component. The 100kb increase is only for uncompressed build outputs. We could optimize this with some build scripts, but the payoff is not worth it right now.

**Comment by larsrickert** [comment]
> Yes, the include happens for every component. The 100kb increase is only for uncompressed build outputs. We could optimize this with some build scripts, but the payoff is not worth it right now.

Hmm I think we should find a way to prevent this. 100kb is a lot in comparisson for the benefits we get when using the layers

**Comment by JoCa96** [comment]
@larsrickert As discussed, added follow-up issue to reduce CSS duplication: https://github.com/SchwarzIT/onyx/issues/812

**Review by larsrickert — DISMISSED** [review]

File: packages/sit-onyx/src/components/OnyxBadge/OnyxBadge.vue [hunk 10]
```diff
@@ -38,7 +38,7 @@ defineSlots<{
 </template>
 
 <style lang="scss">
-@use "../../styles/density.scss";
+@use "../../styles/mixins/density.scss";
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxBadge/OnyxBadge.vue:41]: We also need to use the component layer here, right?

**Review by larsrickert — APPROVED** [review]

File: packages/sit-onyx/src/components/OnyxBadge/OnyxBadge.vue (hunk 10 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxBadge/OnyxBadge.vue:41]: true, done :)

**Review by larsrickert — DISMISSED** [review]

**Review by larsrickert — APPROVED** [review]

**Review by larsrickert — APPROVED** [review]
